trajectory_tracker/src/trajectory_tracker.cpp

```diff
@@ -459,193 +316,77 @@ void TrackerNode::control()
-  float remain_local_ = 0;
-  remain_local_ = dist2d(pos_line_, lpath.poses[iclose].pose.position);
-  for (int i = iclose - 1; i < static_cast<int>(lpath.poses.size()) - 1; i++)
-  {
-    if (i > 2)
-    {
-      geometry_msgs::Point vec = sub2d(lpath.poses[i].pose.position,
-                                       lpath.poses[i - 1].pose.position);
-      float angle = atan2(vec.y, vec.x);
-      float angle_pose_;
-      if (allow_backward_)
-        angle_pose_ = tf2::getYaw(lpath.poses[i + 1].pose.orientation);
-      else
-        angle_pose_ = angle;
-      float signVel_req = cos(angle) * cos(angle_pose_) + sin(angle) * sin(angle_pose_);
-      if (sign_vel_ * signVel_req < 0)
-      {
-        // Stop read forward if the path_ switched back
-        local_goal = i;
-        break;
-      }
-      if (i > iclose)
-        remain_local_ += dist2d(lpath.poses[i - 1].pose.position, lpath.poses[i].pose.position);
-    }
-  }
-  for (int i = iclose - 1; i < local_goal; i++)
-  {
-    if (i > 2)
-    {
-      curv += curv3p(lpath.poses[i - 2].pose.position, lpath.poses[i - 1].pose.position, lpath.poses[i].pose.position);
-    }
-    if (dist2d(lpath.poses[i].pose.position,
-               lpath.poses[local_goal].pose.position) < 0.05)
-      break;
-    if (dist2d(lpath.poses[i].pose.position, pos_line_) > curv_forward_)
-      break;
-  }
-  float remain;
-  remain = dist2d(origin, lpath.poses.back().pose.position);
-  if (min_dist_ < 0 && iclose == local_goal)
-    out_of_line_strip_ = true;
-  if (out_of_line_strip_)
-  {
-    remain = -remain;
-    remain_local_ = -remain_local_;
-  }
-  if (distance_path_ < no_pos_cntl_dist_)
-    remain = remain_local_ = 0;
-  // fprintf(stderr,"%d %d   %0.3f  %+0.3f %+0.3f  %f  %f  sv %f\n",
-  //   out_of_line_strip_, iclose, distance_path_, remain, remain_local_, min_dist_, angle, sign_vel_);
-  // printf("d=%.2f, th=%.2f, curv=%.2f\n", dist, angle, (float)curv);
-  while (angle < -M_PI)
-    angle += 2.0 * M_PI;
-  while (angle > M_PI)
-    angle -= 2.0 * M_PI;
+  const float curv = lpath.getCurvature(it_nearest, it_local_goal, pos_on_line, curv_forward_);
 
   status.distance_remains = remain;
   status.angle_remains = angle;
 
-  float dt = 1.0 / hz_;
-  float _v = v_;
-  float _w = w_;
+  const float dt = 1.0 / hz_;
   // Stop and rotate
-  if ((fabs(rotate_ang_) < M_PI && cos(rotate_ang_) > cos(angle)) ||
-      fabs(remain_local_) < stop_tolerance_dist_ ||
-      distance_path_ < min_track_path_)
+  if ((std::abs(rotate_ang_) < M_PI && std::cos(rotate_ang_) > std::cos(angle)) ||
+      std::abs(remain_local) < stop_tolerance_dist_ ||
+      path_length < min_track_path_ ||
+      in_place_turn_)
   {
-    if (distance_path_ < min_track_path_ || fabs(remain_local_) < stop_tolerance_dist_)
+    if (path_length < min_track_path_ || std::abs(remain_local) < stop_tolerance_dist_)
     {
-      angle = -tf2::getYaw(lpath.poses.back().pose.orientation);
+      angle = -lpath.back().yaw_;
```

Comment:
`it_local_goal->yaw_`
if reached to the local goal, `path_step_done_` should point next to the local goal.